### PR TITLE
Template system implementation

### PR DESCRIPTION
**Pull request by dvoytenko**
Related to #657.

### CODE REVIEW

File: src/template.js [hunk 1]
```diff
@@ -0,0 +1,251 @@
...
+ * The interface that is implemented by all templates.
+ */
+export class BaseTemplate {
+
+  /** @param {!Element} element */
+  constructor(element) {
+    /** @public @const */
+    this.element = element;
+
+    this.compileCallback();
```
Comment (cramforce) [inline src/template.js]: Should this be lazy?

**Comment by cramforce** [comment]
Is it time yet to write a .md file?

File: src/template.js (hunk 1 above)
Comment (dvoytenko) [inline src/template.js]: The templates are always created lazily. So, by proxy, this is lazy as well.

**Comment by dvoytenko** [comment]
@cramforce will add an .md file. good point.

**Comment by dvoytenko** [comment]
@cramforce Added spec MD. And very importantly, added CSS for `template{display:none}` for polyfill.

File: css/amp.css [hunk 2]
```diff
@@ -280,6 +280,12 @@ i-amp-scroll-container {
 }
 
 
+/* Polyfill for IE and any other browser that doesn't understand templates . */
```
Comment (cramforce) [inline css/amp.css]: don't

File: extensions/README.md [hunk 3]
```diff
@@ -24,3 +28,20 @@ Current list of extended components:
 | [`amp-lightbox`](amp-lightbox/amp-lightbox.md)             | Allows for a “lightbox” or similar experience.                                              |
 | [`amp-twitter`](amp-twitter/amp-twitter.md)               | Displays a Twitter Tweet.                                                                   |
 | [`amp-youtube`](amp-youtube/amp-youtube.md)               | Displays a Youtube video.                                                                   |
+
+
+## AMP HTML Extended Templates
```
Comment (cramforce) [inline extensions/README.md]: Maybe note that this is not launched yet?

File: spec/amp-html-format.md [hunk 4]
```diff
@@ -323,12 +323,60 @@ Changes to the PATCH version component (x in 1.1.x) must strictly maintain backw
...
+```
+
+The `<script>` tag must have an `async` attribute and must have a `custom-template` attribute referencing the type of the
+template. The script URL must start with “https://cdn.ampproject.org” and must follow a very strict pattern of
+“/v\d+/[a-z-]+-(latest|\d+|\d+-\d+)\.js”
+
+The templates are declared in the document as following:
+
+```html
+<template type="amp-mustache">
```
Comment (cramforce) [inline spec/amp-html-format.md]: id?

**Comment by cramforce** [comment]
LGTM

File: extensions/README.md (hunk 3 above)
Comment (dvoytenko) [inline extensions/README.md]: Done

File: spec/amp-html-format.md (hunk 4 above)
Comment (dvoytenko) [inline spec/amp-html-format.md]: Done

**Pull request merged**